extensions/windows-event-log/ConsumeWindowsEventLog.cpp

```diff
@@ -248,21 +239,119 @@ void ConsumeWindowsEventLog::onSchedule(const std::shared_ptr<core::ProcessConte
-  if (!subscriptionHandle_) {
-    if (!subscribe(context)) {
-      context->yield();
+void ConsumeWindowsEventLog::processEventsAfterBookmark(core::ProcessSession& session) {
+  // External loop is used in case if there are new events while the events after bookmark are processed.
+  bool hasEvent{};
+  do {
+    hasEvent = false;
+
+    auto hEventResults = EvtQuery(0, wstrChannel_.c_str(), wstrQuery_.c_str(), EvtQueryChannelPath);
+    if (!hEventResults) {
+      logger_->log_error("!EvtQuery error: %d.", GetLastError());
       return;
     }
+    const utils::ScopeGuard guard_hEventResults([hEventResults]() { EvtClose(hEventResults); });
+
+    if (!EvtSeek(hEventResults, 1, pBookmark_->bookmarkHandle(), 0, EvtSeekRelativeToBookmark)) {
+      logger_->log_error("!EvtSeek error: %d.", GetLastError());
+      return;
+    }
+
+    size_t eventCount = 0;
+    auto before_time = std::chrono::high_resolution_clock::now();
+    utils::ScopeGuard timeGuard([&]() {
+      logger_->log_debug("processed %d Events in %llu ms",
+                         eventCount,
+                         std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::high_resolution_clock::now() - before_time).count());
+    });
+
+    // Enumerate the events in the result set after the bookmarked event.
+    bool commitAndSaveBookmark{};
+    std::wstring bookmarkXml;
+    while (true) {
+      EVT_HANDLE hEvent{};
+      DWORD dwReturned{};
+      if (!EvtNext(hEventResults, 1, &hEvent, INFINITE, 0, &dwReturned)) {
+        DWORD status = ERROR_SUCCESS;
+        if (ERROR_NO_MORE_ITEMS != (status = GetLastError())) {
+          logger_->log_error("!EvtNext error %d.", status);
+        }
+        break;
+      }
+      const utils::ScopeGuard guard_hEvent([hEvent]() { EvtClose(hEvent); });
+
+      commitAndSaveBookmark = false;
+      EventRender renderedData;
+      if (processEvent(hEvent, renderedData)) {
+        if (pBookmark_->getNewBookmarkXml(hEvent, bookmarkXml)) {
+          commitAndSaveBookmark = true;
+
+          eventCount++;
+          processEventRender(renderedData, session);
+
+          if (batch_commit_size_ != 0U && (eventCount % batch_commit_size_ == 0)) {
+            auto before_commit = std::chrono::high_resolution_clock::now();
+            session.commit();
+            logger_->log_debug("processQueue commit took %llu ms",
+                               std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::high_resolution_clock::now() - before_commit).count());
+
+            pBookmark_->saveBookmarkXml(bookmarkXml);
+
+            if (session.outgoingConnectionsFull("success")) {
+              return;
+            }
+
+            commitAndSaveBookmark = false;
+          }
+        }
+      }
+
+      hasEvent = true;
+    }
+
+    if (commitAndSaveBookmark) {
+      session.commit();
```

Comment:
Can we postpone `commit()` to outside the outermost loop, running it only once every `onTrigger()` call, unless specified otherwise via `batch_commit_size_`? For performance reasons.

Reply:
Removed outer loop since fixed a bug (should be used handle from last saved xml) `pBookmark_->getBookmarkHandleFromXML())`.